File: 92.Reverse_Linked_List2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode tmp = new ListNode(0);
        for (int i=0; i<m-1; i++) {
            prev = prev.next;
        }
        ListNode curr = prev.next;
        for (int i=0; i<n-m; i++) {
            tmp = prev.next;
            prev.next = curr.next;
            curr.next = curr.next.next;
            prev.next.next =tmp;
        }
        return dummy.next;
    }
}

File: 92.ReverseLinkedList2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        last, curr, prev = head, head, dummy
        for i in range(right):
            if i < left-1:
                prev = prev.next
                curr = curr.next
            last = last.next

        for i in range(right-left+1):
            tmp = curr.next
            curr.next = last
            last, curr = curr, tmp
        prev.next = last
        return dummy.next
